### ISSUE TITLE
The distance between the sample and its nearest negative center is not computed at all.

### ISSUE THREAD

**Issue by MichaelW17** [popcornell/keras-triplet-center-loss#2]
As described in the title, it seems that np.inf is assigned to the intra-class distances in triplet_center_loss in triplet.py, which is more like some kind of placeholder.

**Comment by popcornell** [comment]
If i recall correctly it should be used only to derive the mask. 
Basically by using then reduce_min np.infs are not considered and only the nearest examples should be considered. 
Do you have an example where it fails ?

**Comment by AmberzzZZ** [comment]
The 'y_pred' you send to the function 'triplet_center_loss' contains only the  intra-class distances.
Thus you can not calculate the inter-distance through this function.

**Comment by MichaelW17** [comment]
> If i recall correctly it should be used only to derive the mask.
> Basically by using then reduce_min np.infs are not considered and only the nearest examples should be considered.
> Do you have an example where it fails ?

No it did't fail. Just the "minimus" is supposed to be an L2 distance (between an embedding and its nearest negtive class center), while in the loss function this distance is not computed.

**Comment by MichaelW17** [comment]
So how does the loss func figure out which is the "nearest"?

**Comment by popcornell** [comment]
It is been a while since I don't code in tensorflow/keras having switched to pytorch. 
It seems to me that the triplet_center_loss function takes in input the l2_loss between example and class center and class label:


    model = tf.keras.models.Model(inputs=[x_input, target_input], outputs=[softmax, l2_loss])

    model.compile(loss=["categorical_crossentropy", triplet_center_loss],
                  optimizer=tf.keras.optimizers.Adam(lr=lr), metrics=["accuracy"],
                  loss_weights=loss_weights)`

**Comment by MichaelW17** [comment]
Yeah you right, but notice that only one L2 distance is computed for a certain sample. Triplet center loss requires the distance between a sample and the centers of all classes. 
So, the input to the embedding layer should contain not only one label but all labels, and then index the ground truth embedding in the loss function.

**Comment by popcornell** [comment]
Yes good catch, I think it is more straightforward to have the loss function directly have in input embedding, centroid and label.
